### ISSUE TITLE
Basic Auth to authenticate for token server?

### ISSUE THREAD

**Issue by deitch** [bitly/oauth2_proxy#450]
I have been using oauth2_proxy successfully against a variety of the supported providers, but also against a custom provider just by setting the right vars:

* `redirect_url`
* `login_url`
* `redeem_url`
* `scope`

And a few of the usual others.

The problem I have run into is that oauth2_proxy by default appears to try to authenticate by putting the `client_id` and `client_secret` into the query params rather than using HTTP Basic Auth. See e.g.[Google Redeem()](https://github.com/bitly/oauth2_proxy/blob/master/providers/google.go#L106) and [default Redeem()](https://github.com/bitly/oauth2_proxy/blob/master/providers/provider_default.go#L21), which look like:

```go
	params.Add("client_id", p.ClientID)
	params.Add("client_secret", p.ClientSecret)
``` 

Yet the [RFC specifies](https://tools.ietf.org/html/rfc6749#section-2.3.1) that an oauth2 token server MUST support http basic auth, and that it MAY also support it in params, but is not recommended:

```
The authorization server MUST support the HTTP Basic
   authentication scheme for authenticating clients that were issued a
   client password.

   For example (with extra line breaks for display purposes only):

     Authorization: Basic czZCaGRSa3F0Mzo3RmpmcDBaQnIxS3REUmJuZlZkbUl3

   Alternatively, the authorization server MAY support including the
   client credentials in the request-body using the following
   parameters:

   client_id
         REQUIRED.  The client identifier issued to the client during
         the registration process described by Section 2.2.

   client_secret
         REQUIRED.  The client secret.  The client MAY omit the
         parameter if the client secret is an empty string.




Hardt                        Standards Track                   [Page 16]
 
RFC 6749                        OAuth 2.0                   October 2012


   Including the client credentials in the request-body using the two
   parameters is NOT RECOMMENDED and SHOULD be limited to clients unable
   to directly utilize the HTTP Basic authentication scheme (or other
   password-based HTTP authentication schemes).  The parameters can only
   be transmitted in the request-body and MUST NOT be included in the
   request URI.
```

Many implementations only support HTTP Basic.

1. Is this supported somewhere and I missed it?
2. If not, the change looks simple, would a PR be accepted?

**Comment by jehiah** [comment]
@deitch this is similar to #392. I'd be happy to accept PR's to update any for providers that do support `HTTP Basic`.

**Comment by deitch** [comment]
Yeah, it is similar. My search parameters were more for "Basic" than "Authorization". Oops.

I don't know who uses HTTP Basic Auth, probably could look. I more use it for custom providers, just pick any provider and override the elements above. 

So.. if we find a provider that supports it (all should, but...) then we just change it there; if not, we make it an option?

**Comment by deitch** [comment]
Nothing anywhere in Google Docs that says they do support it. Nothing says they _don't_, but nothing says they do. How odd, since they usually are quite good about following the RFCs.

**Comment by deitch** [comment]
I can confirm that GitHub supports http basic auth (as it should). Will check Google now.

**Comment by deitch** [comment]
In the meantime, do we want it as an _option_ to use basic auth, or the _default_ with GitHub and Google and others?

**Comment by ploxiln** [comment]
I don't think we want to add options for this. I think we want client-id/client-secret to be passed as http basic auth for any providers it will work for.

**Comment by deitch** [comment]
OK. 

Truth is, they _all_ should support it, since the RFC says "MUST". But _should_ and _do_ are a far way away.

I am going to start with GitHub and Google, implement in there, test, and get a PR going on those. We can expand over time.

**Comment by deitch** [comment]
I took a slightly different approach. Every provider except Google and oidc uses the default `Redeem()`. So instead of implementing in each, I added `HttpBasicAuth bool` to the `ProviderData` with a default of `false`. 

As soon as a we know a provider supports it, all we need to do is set that parameter. PR coming soon.

**Comment by ploxiln** [comment]
actually we use "gpm": https://github.com/pote/gpm

**Comment by deitch** [comment]
PR opened for first part in #451 . Once that is in, I will add some providers one-by-one by enabling the flag and testing it. I already know that GitHub supports it. Truth is, according to the RFC, _everyone_ should support it, but "should" and "do" often are far away... :-)

**Comment by ploxiln** [comment]
I see now, via #451, that `client_id` and `client_secret` were being put in the body of a POST request, not in url query params, which is the impression I got from your original report. Thus, this is not really a significant improvement for security or safety.

This change might still be desired in order to more closely match the oauth2 rfc and possibly be more compatible with other future implementations of it. But it fights against the historical inertia of all the current providers working with and being tested and validated with `client_secret` in the POST body.

**Comment by deitch** [comment]
Right you are. I had glossed over it the first time and in seeing it URL-encoded thought it was in the query string. Only once starting the PR did I realize that it was encoded but in the body. 

Either way, the RFC provides it as an option in the body but a requirement in http basic auth. 

I actually had offered to put it as a runtime option, but someone (on phone now, sorry cannot check) said it is better as a provider setting. 

Next steps?

**Comment by ploxiln** [comment]
That was me :) I still think it doesn't make sense as a user-set option, since each provider has custom code anyway - it should know what the best option is for that provider.

I want to wait for a couple more opinions on any possible benefits or drawbacks of using basic auth for `client_secret` (now that we know it's currently passed in a POST body).

Just BTW, (only) the google provider overrides the default `Redeem()` method.

**Comment by deitch** [comment]
> That was me :)

Oh. Oops, sorry. Was on my phone.

> I still think it doesn't make sense as a user-set option, since each provider has custom code anyway - it should know what the best option is for that provider.

I agreed, so I did it this way.

> want to wait for a couple more opinions on any possible benefits or drawbacks of using basic auth for client_secret (now that we know it's currently passed in a POST body).

Sure, makes sense. I still think the _default_ should be basic auth and fallback be body for the secret, per the RFC, but every provider and its idiosyncrasies.

> (only) the google provider overrides the default Redeem() method.

Saw that too. I didn't dig too deeply into why, will once this is approved and merged and we can start one-by-one enabling.

**Comment by deitch** [comment]
@jehiah still around? Can we get this through (or changes if necessary), and then I can start opening PRs for individual providers after testing?
